aten/src/ATen/native/Linear.cpp

```diff
@@ -136,241 +136,319 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
-        }
-      } else {                                          // a letter (hopefully)
-        TORCH_CHECK((ell_char_count == 0) || (ell_char_count == 3), "'.' must only occur in ellipsis, operand ", operand);
-        TORCH_CHECK(('a' <= c) && (c <= 'z'), "only lowercase letters a-z allowed as indices");
-        int64_t letter_num = c-'a';                     // letter_num  = position in letter_mapping
-        if (letter_mapping[letter_num] == -1) {         // new letter, add internal index and mapping
-          letter_mapping[letter_num] = num_total_idxes;
-          num_total_idxes++;
-          last_idx_occurrence.push_back(operand);
-        } else {                                        // letter we have already seen
-          last_idx_occurrence[letter_mapping[letter_num]] = operand;
-        }
-        num_letter_occurrences[letter_num]++;
-        current_op_idxes.push_back(letter_mapping[letter_num]);
-        dims_in_term++;
-      }
+// There are roughly three parts to compute einsum:
+// 1. Parse equation to extract the labels for each input operand and output
+// 2. Unsqueeze missing dimensions from input operands and permute to align them
+// 3. Compute result by multiplying input operands and summing contraction
+//    dimensions We do the last part by reducing to bmm, there is room for
+//    optimization here.
+Tensor einsum(std::string equation, TensorList operands) {
+  TORCH_CHECK(!operands.empty(), "einsum() must provide at least one operand");
+  checkDeviceType("einsum()", operands, operands[0].device().type());
+
+  // Code for encoding ellipsis with labels (...)
+  constexpr int ELLIPSIS = '.';
+
+  // Find arrow (->) to split equation into lhs and rhs
+  std::size_t arrow_pos = equation.find("->");
+
+  // Parse lhs of equation to store ASCII value for each valid label and
+  // ELLIPSIS for every operand into op_labels
+  std::string lhs = equation.substr(0, arrow_pos);
+  std::vector<std::vector<int>> op_labels(operands.size());
+  bool found_ell = false;
+  std::size_t curr_op = 0;
+  for (std::size_t i = 0; i < lhs.length(); ++i) {
+    switch (lhs[i]) {
+      case ' ':
+        // Ignore spaces
+        break;
+
+      case '.':
+        TORCH_CHECK(
+            // Only one ellipsis per operand can be given
+            !found_ell,
+            "einsum() found \'.\' for operand ",
+            curr_op,
+            " for which an ellipsis was already found");
+        TORCH_CHECK(
+            // Ensure it's a valid ellipsis
+            i + 2 < lhs.length() && lhs[++i] == '.' && lhs[++i] == '.',
+            "einsum() found \'.\' for operand ",
+            curr_op,
+            " that is not part of any ellipsis");
+        op_labels[curr_op].push_back(ELLIPSIS);
+        found_ell = true;
+        break;
+
+      case ',':
+        // Move onto next operand
+        TORCH_CHECK(
+            ++curr_op < operands.size(),
```

Comment:
Separate the increment to curr_op from the TORCH_CHECK to make this more readable.